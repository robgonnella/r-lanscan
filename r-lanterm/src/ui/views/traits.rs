use std::sync::mpsc::Sender;

use ratatui::{crossterm::event::Event as CrossTermEvent, layout::Rect};

use crate::ui::{
    events::types::Event,
    store::state::{State, ViewID},
};

pub trait EventHandler {
    fn process_event(&self, evt: &CrossTermEvent, ctx: &CustomWidgetContext) -> bool;
}

pub struct CustomWidgetContext {
    // app state
    pub state: State,
    // total area for the entire application - useful for calculating
    // popover areas
    pub app_area: Rect,
    // event producer - this how components and views can communicate user
    // behavior back to main loop to perform actions that aren't related to
    // state - executing a shell command
    pub events: Sender<Event>,
}

pub trait CustomWidget {
    fn render(self, area: Rect, buf: &mut ratatui::prelude::Buffer, ctx: &CustomWidgetContext);
}

pub trait CustomWidgetRef {
    fn render_ref(&self, area: Rect, buf: &mut ratatui::prelude::Buffer, ctx: &CustomWidgetContext);
}

pub trait CustomStatefulWidget {
    type State;

    fn render(
        self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        state: &mut Self::State,
        ctx: &CustomWidgetContext,
    );
}

pub trait View: EventHandler + CustomWidgetRef {
    fn id(&self) -> ViewID;
    fn legend(&self, _state: &State) -> &str {
        ""
    }
    fn override_main_legend(&self, _state: &State) -> bool {
        false
    }
}
