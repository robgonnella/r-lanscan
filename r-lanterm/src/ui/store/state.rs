use core::fmt;
use std::{collections::HashMap, process::Output};

use r_lanlib::scanners::{Device, DeviceWithPorts};

use crate::{
    config::{Config, DeviceConfig},
    ui::{
        colors::{Colors, Theme},
        events::types::Command,
    },
};

pub type MissedCount = i8;

#[derive(Debug, Clone, Eq, Hash, PartialEq)]
pub enum ViewID {
    Main,
    Device,
    Devices,
    Config,
    ViewSelect,
}

impl fmt::Display for ViewID {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[derive(Debug, Clone)]
pub struct State {
    pub true_color_enabled: bool,
    pub ui_paused: bool,
    pub error: Option<String>,
    pub render_view_select: bool,
    pub view_id: ViewID,
    pub config: Config,
    pub arp_history: HashMap<String, (Device, MissedCount)>,
    pub devices: Vec<DeviceWithPorts>,
    pub device_map: HashMap<String, DeviceWithPorts>,
    pub selected_device: Option<DeviceWithPorts>,
    pub selected_device_config: Option<DeviceConfig>,
    pub colors: Colors,
    pub message: Option<String>,
    pub cmd_in_progress: Option<Command>,
    pub cmd_output: Option<(Command, Output)>,
}

#[cfg(test)]
impl State {
    pub fn default() -> Self {
        let config = Config::default();
        let theme = Theme::from_string(&config.theme);
        let true_color_enabled = true;
        let colors = crate::ui::colors::Colors::new(
            theme.to_palette(true_color_enabled),
            true_color_enabled,
        );

        Self {
            true_color_enabled,
            ui_paused: false,
            error: None,
            render_view_select: false,
            view_id: ViewID::Devices,
            config,
            arp_history: HashMap::new(),
            devices: Vec::new(),
            device_map: HashMap::new(),
            selected_device: None,
            selected_device_config: None,
            colors,
            message: None,
            cmd_in_progress: None,
            cmd_output: None,
        }
    }
}
