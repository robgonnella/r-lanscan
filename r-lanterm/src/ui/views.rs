pub mod config;
pub mod device;
pub mod devices;
pub mod main;
pub mod traits;
pub mod view_select;
