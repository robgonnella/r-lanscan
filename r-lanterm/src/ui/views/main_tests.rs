use insta::assert_snapshot;
use nanoid::nanoid;
use pnet::util::MacAddr;
use r_lanlib::scanners::{DeviceWithPorts, Port};
use ratatui::{backend::TestBackend, Terminal};
use std::{
    collections::HashSet,
    fs,
    sync::{mpsc, Mutex},
};

use crate::config::{Config, ConfigManager};

use super::*;

fn setup() -> (MainView, Arc<Store>, String) {
    fs::create_dir_all("generated").unwrap();
    let tmp_path = format!("generated/{}.yml", nanoid!());
    let conf_manager = Arc::new(Mutex::new(ConfigManager::new(tmp_path.as_str())));
    let store = Arc::new(Store::new(conf_manager));
    let config = Config {
        id: "default".to_string(),
        cidr: "192.168.1.1/24".to_string(),
        default_ssh_identity: "id_rsa".to_string(),
        default_ssh_port: "22".to_string(),
        default_ssh_user: "user".to_string(),
        device_configs: HashMap::new(),
        ports: vec![],
        theme: "Blue".to_string(),
    };
    store.dispatch(Action::CreateAndSetConfig(config));

    let mut open_ports: HashSet<Port> = HashSet::new();
    open_ports.insert(Port {
        id: 80,
        service: "http".to_string(),
    });

    let device_1 = DeviceWithPorts {
        hostname: "hostname".to_string(),
        ip: "10.10.10.1".to_string(),
        is_current_host: false,
        mac: MacAddr::default().to_string(),
        open_ports: open_ports.clone(),
        vendor: "mac".to_string(),
    };

    let device_2 = DeviceWithPorts {
        hostname: "dev2_hostname".to_string(),
        ip: "10.10.10.2".to_string(),
        is_current_host: true,
        mac: "ff:ff:ff:ff:ff:ff".to_string(),
        open_ports,
        vendor: "linux".to_string(),
    };

    store.dispatch(Action::AddDevice(device_1.clone()));
    store.dispatch(Action::AddDevice(device_2.clone()));
    let (tx, _rx) = mpsc::channel();
    (MainView::new(Arc::clone(&store), tx), store, tmp_path)
}

fn tear_down(conf_path: String) {
    fs::remove_file(conf_path).unwrap();
}

#[test]
fn test_main_view() {
    let (main_view, store, conf_path) = setup();
    let mut terminal = Terminal::new(TestBackend::new(80, 15)).unwrap();
    let state = store.get_state();
    let channel = std::sync::mpsc::channel();

    terminal
        .draw(|frame| {
            let ctx = CustomWidgetContext {
                state,
                app_area: frame.area(),
                events: channel.0,
            };

            main_view.render_ref(frame.area(), frame.buffer_mut(), &ctx);
        })
        .unwrap();

    assert_snapshot!(terminal.backend());
    tear_down(conf_path);
}
