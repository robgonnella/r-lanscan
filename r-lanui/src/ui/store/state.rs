use core::fmt;
use std::{collections::HashMap, fmt::Display, process::Output};

use r_lanlib::scanners::DeviceWithPorts;
use ratatui::style::{palette::tailwind, Color};

use crate::config::{Config, DeviceConfig};

#[derive(Debug, Clone, Eq, Hash, PartialEq)]
pub enum ViewID {
    Main,
    Device,
    Devices,
    Config,
    ViewSelect,
}

impl fmt::Display for ViewID {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[derive(Clone, Debug)]
pub struct Colors {
    pub buffer_bg: Color,
    pub header_bg: Color,
    pub header_fg: Color,
    pub selected_row_fg: Color,
    pub row_fg: Color,
    pub row_bg: Color,
    pub border_color: Color,
    pub scroll_bar_fg: Color,
    pub label: Color,
}

impl Colors {
    pub fn new(color: &tailwind::Palette) -> Self {
        Self {
            buffer_bg: tailwind::SLATE.c950,
            header_bg: color.c900,
            header_fg: tailwind::SLATE.c200,
            selected_row_fg: color.c400,
            row_fg: tailwind::SLATE.c200,
            row_bg: tailwind::SLATE.c950,
            border_color: color.c400,
            scroll_bar_fg: tailwind::SLATE.c800,
            label: color.c400,
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Command {
    SSH,
    TRACEROUTE,
}

impl Display for Command {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Command::SSH => write!(f, "ssh"),
            Command::TRACEROUTE => write!(f, "traceroute"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct State {
    pub render_view_select: bool,
    pub view_id: ViewID,
    pub config: Config,
    pub devices: Vec<DeviceWithPorts>,
    pub device_map: HashMap<String, DeviceWithPorts>,
    pub selected_device: Option<DeviceWithPorts>,
    pub selected_device_config: Option<DeviceConfig>,
    pub colors: Colors,
    pub message: Option<String>,
    pub execute_cmd: Option<Command>,
    pub cmd_output: Option<(Command, Output)>,
}

#[derive(Debug, Eq, PartialEq, Clone)]
pub enum Theme {
    Blue,
    Emerald,
    Indigo,
    Red,
}

impl Theme {
    pub fn from_string(value: &String) -> Theme {
        match value.as_str() {
            "Blue" => Theme::Blue,
            "Emerald" => Theme::Emerald,
            "Indigo" => Theme::Indigo,
            "Red" => Theme::Red,
            _ => Theme::Blue,
        }
    }

    pub fn to_string(&self) -> String {
        match self {
            Theme::Blue => "Blue".to_string(),
            Theme::Emerald => "Emerald".to_string(),
            Theme::Indigo => "Indigo".to_string(),
            Theme::Red => "Red".to_string(),
        }
    }

    pub fn to_palette(&self) -> &'static tailwind::Palette {
        match self {
            Theme::Blue => &tailwind::BLUE,
            Theme::Emerald => &tailwind::EMERALD,
            Theme::Indigo => &tailwind::INDIGO,
            Theme::Red => &tailwind::RED,
        }
    }
}
