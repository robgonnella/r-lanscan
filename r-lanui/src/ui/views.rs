use ratatui::{crossterm::event::Event, Frame};

pub mod config;
pub mod device;
pub mod devices;

pub trait View {
    fn render(&mut self, f: &mut Frame);
    fn process_event(&mut self, evt: &Event) -> bool;
}
