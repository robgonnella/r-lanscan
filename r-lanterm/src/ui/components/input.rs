use ratatui::{
    layout::Rect,
    style::Style,
    text::{Line, Span},
    widgets::Widget,
};

use crate::ui::views::traits::{CustomStatefulWidget, CustomWidgetContext};

#[derive(Debug, Clone)]
pub struct InputState {
    pub editing: bool,
    pub value: String,
}

pub struct Input {
    label: String,
}

impl Input {
    pub fn new(label: &str) -> Self {
        Self {
            label: String::from(label),
        }
    }
}

impl CustomStatefulWidget for Input {
    type State = InputState;

    fn render(
        self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        state: &mut Self::State,
        ctx: &CustomWidgetContext,
    ) where
        Self: Sized,
    {
        let label = Span::from(format!("{0}: ", self.label));
        let mut style = Style::default();
        if state.editing {
            style = style.fg(ctx.state.colors.input_editing);
        }
        let value = Span::from(state.value.as_str()).style(style);
        let line = Line::from(vec![label, value]);
        line.render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use crate::ui::store::state::State;

    use super::*;
    use insta::assert_snapshot;
    use ratatui::{backend::TestBackend, Terminal};

    #[test]
    fn renders_input_component_non_edit_mode() {
        let input = Input::new("Test");

        let mut input_state = InputState {
            editing: false,
            value: "value".to_string(),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 3)).unwrap();
        let state = State::default();
        let channel = std::sync::mpsc::channel();

        terminal
            .draw(|frame| {
                let ctx = CustomWidgetContext {
                    state,
                    app_area: frame.area(),
                    events: channel.0,
                };

                input.render(frame.area(), frame.buffer_mut(), &mut input_state, &ctx);
            })
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn renders_input_component_edit_mode() {
        let input = Input::new("Test");

        let mut input_state = InputState {
            editing: true,
            value: "value".to_string(),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 3)).unwrap();
        let state = State::default();
        let channel = std::sync::mpsc::channel();

        terminal
            .draw(|frame| {
                let ctx = CustomWidgetContext {
                    state,
                    app_area: frame.area(),
                    events: channel.0,
                };

                input.render(frame.area(), frame.buffer_mut(), &mut input_state, &ctx);
            })
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
