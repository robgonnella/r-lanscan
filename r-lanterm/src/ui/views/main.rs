use std::{
    collections::HashMap,
    rc::Rc,
    sync::{mpsc::Sender, Arc},
};

use crate::ui::{
    components::{footer::InfoFooter, header::Header, popover::get_popover_area},
    events::types::Event,
    store::{
        action::Action,
        state::{State, ViewID},
        store::Store,
    },
};
use ratatui::{
    crossterm::event::{Event as CrossTermEvent, KeyCode},
    layout::{Constraint, Layout, Rect},
    style::{palette::tailwind, Style},
    text::Line,
    widgets::{Block, BorderType, Clear as ClearWidget, Padding, Paragraph, Widget, WidgetRef},
};

use super::{
    config::ConfigView,
    device::DeviceView,
    devices::DevicesView,
    traits::{CustomWidget, CustomWidgetContext, CustomWidgetRef, EventHandler, View},
    view_select::ViewSelect,
};

const DEFAULT_PADDING: Padding = Padding::horizontal(2);

pub struct MainView {
    store: Arc<Store>,
    sub_views: HashMap<ViewID, Box<dyn View>>,
    _tx: Sender<Event>,
}

impl MainView {
    pub fn new(store: Arc<Store>, tx: Sender<Event>) -> Self {
        let mut sub_views: HashMap<ViewID, Box<dyn View>> = HashMap::new();

        let config = Box::new(ConfigView::new(Arc::clone(&store)));
        let device = Box::new(DeviceView::new(Arc::clone(&store)));
        let devices = Box::new(DevicesView::new(Arc::clone(&store)));
        let view_select = Box::new(ViewSelect::new(
            vec![ViewID::Devices, ViewID::Config],
            2,
            Arc::clone(&store),
        ));

        sub_views.insert(config.id(), config);
        sub_views.insert(device.id(), device);
        sub_views.insert(devices.id(), devices);
        sub_views.insert(view_select.id(), view_select);

        Self {
            store,
            sub_views,
            _tx: tx,
        }
    }

    fn render_buffer_bg(&self, area: Rect, buf: &mut ratatui::prelude::Buffer, state: &State) {
        let block = Block::new()
            .style(Style::new().bg(state.colors.buffer_bg))
            .padding(DEFAULT_PADDING);
        block.render(area, buf);
    }

    fn get_top_section_areas(&self, area: Rect) -> Rc<[Rect]> {
        Layout::horizontal([
            Constraint::Percentage(20),
            Constraint::Percentage(100),
            Constraint::Percentage(20),
        ])
        .split(area)
    }

    fn render_top(
        &self,
        sections: Rc<[Rect]>,
        buf: &mut ratatui::prelude::Buffer,
        message: Option<String>,
        ctx: &CustomWidgetContext,
    ) {
        let logo =
            Paragraph::new("\nr-lanui").style(Style::new().fg(ctx.state.colors.border_color));
        let logo_block: Block<'_> = Block::bordered()
            .border_style(Style::new().fg(ctx.state.colors.border_color))
            .border_type(BorderType::Double)
            .padding(DEFAULT_PADDING);
        let logo_inner_area = logo_block.inner(sections[0]);

        logo_block.render(sections[0], buf);
        logo.render_ref(logo_inner_area, buf);

        if let Some(message) = message {
            let message_block = Block::default().padding(Padding::uniform(2));
            let message_inner_area = message_block.inner(sections[1]);
            let m = Header::new(format!("\n\n{message}"));
            message_block.render(sections[1], buf);
            m.render(message_inner_area, buf, ctx);
        }

        let current_view = Paragraph::new(format!("\n{} â–¼", ctx.state.view_id.to_string()))
            .style(Style::new().fg(ctx.state.colors.border_color));
        let current_view_block = Block::bordered()
            .border_style(Style::new().fg(ctx.state.colors.border_color))
            .border_type(BorderType::Double)
            .padding(DEFAULT_PADDING);
        let current_view_inner_area = current_view_block.inner(sections[2]);

        current_view_block.render(sections[2], buf);
        current_view.render_ref(current_view_inner_area, buf);
    }

    fn render_middle_view(
        &self,
        view: &Box<dyn View>,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        let block: Block<'_> = Block::bordered()
            .border_style(Style::new().fg(ctx.state.colors.border_color))
            .border_type(BorderType::Plain)
            .padding(DEFAULT_PADDING);
        let inner_area = block.inner(area);
        block.render(area, buf);
        view.render_ref(inner_area, buf, ctx);
    }

    fn render_view_select_popover(
        &self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        let view = self.sub_views.get(&ViewID::ViewSelect);

        if let Some(view_select) = view {
            view_select.render_ref(area, buf, ctx);
        }
    }

    fn render_error_popover(&self, area: Rect, buf: &mut ratatui::prelude::Buffer, state: &State) {
        if state.error.is_some() {
            let msg = state.error.clone().unwrap();
            let block = Block::bordered()
                .border_type(BorderType::Double)
                .border_style(
                    Style::new()
                        .fg(tailwind::RED.c600)
                        .bg(state.colors.buffer_bg),
                )
                .padding(Padding::uniform(2))
                .style(Style::default().bg(state.colors.buffer_bg));
            let inner_area = block.inner(area);
            let [msg_area, exit_area] = Layout::vertical([
                Constraint::Percentage(100), // msg
                Constraint::Length(1),       // exit
            ])
            .areas(inner_area);

            let message = Line::from(format!("Error: {}", msg));
            let exit = Paragraph::new("Press enter to clear error").centered();
            ClearWidget.render(area, buf);
            block.render(area, buf);
            message.render(msg_area, buf);
            exit.render(exit_area, buf);
        }
    }

    fn render_footer(
        &self,
        legend: &str,
        override_legend: bool,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        if override_legend {
            let footer = InfoFooter::new(legend.to_string());
            footer.render(area, buf, ctx);
        } else {
            let mut info = String::from("(q) quit | (v) change view");

            if legend.len() > 0 {
                info = format!("{info} | {legend}");
            }

            let footer = InfoFooter::new(info);
            footer.render(area, buf, ctx);
        }
    }
}

impl View for MainView {
    fn id(&self) -> ViewID {
        ViewID::Main
    }
}

impl CustomWidgetRef for MainView {
    fn render_ref(
        &self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        // consists of 3 vertical rectangles (top, middle, bottom)
        let page_areas = Layout::vertical([
            Constraint::Length(5),
            Constraint::Min(10),
            Constraint::Length(3),
        ])
        .split(area);

        let view_id = ctx.state.view_id.clone();
        let view = self.sub_views.get(&view_id).unwrap();
        let legend = view.legend(&ctx.state);
        let override_legend = view.override_main_legend(&ctx.state);

        // render background for entire display
        self.render_buffer_bg(area, buf, &ctx.state);
        // logo & view select
        let top_section_areas = self.get_top_section_areas(page_areas[0]);
        let top_section_areas_clone = Rc::clone(&top_section_areas);
        self.render_top(top_section_areas, buf, ctx.state.message.clone(), ctx);
        // view
        self.render_middle_view(view, page_areas[1], buf, ctx);
        // legend for current view
        self.render_footer(legend, override_legend, page_areas[2], buf, ctx);

        // view selection
        if ctx.state.render_view_select {
            let mut select_area = top_section_areas_clone[2];
            select_area.height = (self.sub_views.len() * 3).try_into().unwrap();

            let select_block = Block::bordered()
                .border_style(Style::new().fg(ctx.state.colors.border_color))
                .border_type(BorderType::Double);

            let select_inner_area = select_block.inner(select_area);

            select_block.render(select_area, buf);

            ClearWidget.render(select_inner_area, buf);
            self.render_buffer_bg(select_inner_area, buf, &ctx.state);
            self.render_view_select_popover(select_inner_area, buf, ctx);
        }

        // popover when there are errors in the store
        // important to render this last so it properly layers on top
        self.render_error_popover(get_popover_area(area, 50, 40), buf, &ctx.state);
    }
}

impl EventHandler for MainView {
    fn process_event(&self, evt: &CrossTermEvent, ctx: &CustomWidgetContext) -> bool {
        if ctx.state.render_view_select {
            let select_view = self.sub_views.get(&ViewID::ViewSelect).unwrap();
            return select_view.process_event(evt, ctx);
        }

        if ctx.state.error.is_some() {
            match evt {
                CrossTermEvent::Key(key) => match key.code {
                    KeyCode::Enter => {
                        self.store.dispatch(Action::SetError(None));
                    }
                    _ => {}
                },
                _ => {}
            }
            true
        } else {
            let view_id = ctx.state.view_id.clone();
            let view = self.sub_views.get(&view_id).unwrap();
            let mut handled = view.process_event(evt, ctx);

            if !handled {
                match evt {
                    CrossTermEvent::Key(key) => match key.code {
                        KeyCode::Char('v') => {
                            if !ctx.state.render_view_select {
                                handled = true;
                                self.store.dispatch(Action::ToggleViewSelect);
                            }
                        }
                        KeyCode::Esc => {
                            if ctx.state.render_view_select {
                                handled = true;
                                self.store.dispatch(Action::ToggleViewSelect);
                            }
                        }
                        _ => {}
                    },
                    _ => {}
                }
            }

            handled
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use nanoid::nanoid;
    use pnet::util::MacAddr;
    use r_lanlib::scanners::{DeviceWithPorts, Port};
    use ratatui::{backend::TestBackend, Terminal};
    use std::{
        collections::HashSet,
        fs,
        sync::{mpsc, Mutex},
    };

    use crate::config::{Config, ConfigManager};

    use super::*;

    fn setup() -> (MainView, Arc<Store>, String) {
        fs::create_dir_all("generated").unwrap();
        let tmp_path = format!("generated/{}.yml", nanoid!());
        let conf_manager = Arc::new(Mutex::new(ConfigManager::new(tmp_path.as_str())));
        let store = Arc::new(Store::new(conf_manager));
        let config = Config {
            id: "default".to_string(),
            cidr: "192.168.1.1/24".to_string(),
            default_ssh_identity: "id_rsa".to_string(),
            default_ssh_port: "22".to_string(),
            default_ssh_user: "user".to_string(),
            device_configs: HashMap::new(),
            ports: vec![],
            theme: "Blue".to_string(),
        };
        store.dispatch(Action::CreateAndSetConfig(config));

        let mut open_ports: HashSet<Port> = HashSet::new();
        open_ports.insert(Port {
            id: 80,
            service: "http".to_string(),
        });

        let device_1 = DeviceWithPorts {
            hostname: "hostname".to_string(),
            ip: "10.10.10.1".to_string(),
            is_current_host: false,
            mac: MacAddr::default().to_string(),
            open_ports: open_ports.clone(),
            vendor: "mac".to_string(),
        };

        let device_2 = DeviceWithPorts {
            hostname: "dev2_hostname".to_string(),
            ip: "10.10.10.2".to_string(),
            is_current_host: true,
            mac: "ff:ff:ff:ff:ff:ff".to_string(),
            open_ports,
            vendor: "linux".to_string(),
        };

        store.dispatch(Action::AddDevice(device_1.clone()));
        store.dispatch(Action::AddDevice(device_2.clone()));
        let (tx, _rx) = mpsc::channel();
        (MainView::new(Arc::clone(&store), tx), store, tmp_path)
    }

    fn tear_down(conf_path: String) {
        fs::remove_file(conf_path).unwrap();
    }

    #[test]
    fn test_main_view() {
        let (main_view, store, conf_path) = setup();
        let mut terminal = Terminal::new(TestBackend::new(80, 15)).unwrap();
        let state = store.get_state();
        let channel = std::sync::mpsc::channel();

        terminal
            .draw(|frame| {
                let ctx = CustomWidgetContext {
                    state,
                    app_area: frame.area(),
                    events: channel.0,
                };

                main_view.render_ref(frame.area(), frame.buffer_mut(), &ctx);
            })
            .unwrap();

        assert_snapshot!(terminal.backend());
        tear_down(conf_path);
    }
}
