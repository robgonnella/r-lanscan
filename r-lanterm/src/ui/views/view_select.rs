use std::{cell::RefCell, sync::Arc};

use itertools::Itertools;
use ratatui::crossterm::event::{Event, KeyCode, KeyEventKind};

use crate::ui::{
    components::table::{self, Table},
    store::{action::Action, state::ViewID, store::Store},
};

use super::traits::{CustomWidgetContext, CustomWidgetRef, EventHandler, View};

pub struct ViewSelect {
    store: Arc<Store>,
    view_ids: Vec<ViewID>,
    table: RefCell<Table>,
}

impl ViewSelect {
    pub fn new(view_ids: Vec<ViewID>, padding: usize, store: Arc<Store>) -> Self {
        let mut spacer = String::from("");

        if padding > 0 {
            for _ in 0..padding {
                spacer += " ";
            }
        }

        let table_items = view_ids
            .clone()
            .iter()
            .map(|v| vec![format!("{}{}", spacer, v.to_string())])
            .collect_vec();

        let mut table_select = Table::new(
            table_items,
            None,
            vec![15; view_ids.len()],
            table::DEFAULT_ITEM_HEIGHT,
        );

        table_select.next();

        Self {
            store,
            view_ids,
            table: RefCell::new(table_select),
        }
    }

    fn next(&self) {
        self.table.borrow_mut().next();
    }

    fn previous(&self) {
        self.table.borrow_mut().previous();
    }

    fn handle_selected(&self) {
        let i = self.table.borrow().selected();
        if let Some(selected) = i {
            let id = self.view_ids[selected].clone();
            self.store.dispatch(Action::UpdateView(id));
            self.store.dispatch(Action::ToggleViewSelect);
        }
    }
}

impl View for ViewSelect {
    fn id(&self) -> ViewID {
        ViewID::ViewSelect
    }
}

impl EventHandler for ViewSelect {
    fn process_event(&self, evt: &Event, ctx: &CustomWidgetContext) -> bool {
        if !ctx.state.render_view_select {
            return false;
        }

        let mut handled = false;

        match evt {
            Event::Key(key) => {
                if key.kind == KeyEventKind::Press {
                    match key.code {
                        KeyCode::Char('j') | KeyCode::Down => {
                            self.next();
                            handled = true;
                        }
                        KeyCode::Char('k') | KeyCode::Up => {
                            self.previous();
                            handled = true;
                        }
                        KeyCode::Esc => {
                            if ctx.state.render_view_select {
                                self.store.dispatch(Action::ToggleViewSelect);
                                handled = true;
                            }
                        }
                        KeyCode::Enter => {
                            self.handle_selected();
                            handled = true;
                        }
                        _ => {}
                    }
                }
            }
            _ => {}
        }

        handled
    }
}

impl CustomWidgetRef for ViewSelect {
    fn render_ref(
        &self,
        area: ratatui::prelude::Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        self.table.borrow().render_ref(area, buf, ctx);
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use nanoid::nanoid;
    use pnet::util::MacAddr;
    use r_lanlib::scanners::{DeviceWithPorts, Port};
    use ratatui::{backend::TestBackend, Terminal};
    use std::{
        collections::{HashMap, HashSet},
        fs,
        sync::Mutex,
    };

    use crate::config::{Config, ConfigManager};

    use super::*;

    fn setup() -> (ViewSelect, Arc<Store>, String) {
        fs::create_dir_all("generated").unwrap();
        let tmp_path = format!("generated/{}.yml", nanoid!());
        let conf_manager = Arc::new(Mutex::new(ConfigManager::new(tmp_path.as_str())));
        let store = Arc::new(Store::new(conf_manager));
        let config = Config {
            id: "default".to_string(),
            cidr: "192.168.1.1/24".to_string(),
            default_ssh_identity: "id_rsa".to_string(),
            default_ssh_port: "22".to_string(),
            default_ssh_user: "user".to_string(),
            device_configs: HashMap::new(),
            ports: vec![],
            theme: "Blue".to_string(),
        };
        store.dispatch(Action::CreateAndSetConfig(config));

        let mut open_ports: HashSet<Port> = HashSet::new();
        open_ports.insert(Port {
            id: 80,
            service: "http".to_string(),
        });

        let device_1 = DeviceWithPorts {
            hostname: "hostname".to_string(),
            ip: "10.10.10.1".to_string(),
            is_current_host: false,
            mac: MacAddr::default().to_string(),
            open_ports: open_ports.clone(),
            vendor: "mac".to_string(),
        };

        let device_2 = DeviceWithPorts {
            hostname: "dev2_hostname".to_string(),
            ip: "10.10.10.2".to_string(),
            is_current_host: true,
            mac: "ff:ff:ff:ff:ff:ff".to_string(),
            open_ports,
            vendor: "linux".to_string(),
        };

        store.dispatch(Action::AddDevice(device_1.clone()));
        store.dispatch(Action::AddDevice(device_2.clone()));
        let view_ids = vec![ViewID::Devices, ViewID::Config];
        (
            ViewSelect::new(view_ids, 2, Arc::clone(&store)),
            store,
            tmp_path,
        )
    }

    fn tear_down(conf_path: String) {
        fs::remove_file(conf_path).unwrap();
    }

    #[test]
    fn test_view_select_view() {
        let (view_select, store, conf_path) = setup();
        let mut terminal = Terminal::new(TestBackend::new(80, 15)).unwrap();
        let state = store.get_state();
        let channel = std::sync::mpsc::channel();

        terminal
            .draw(|frame| {
                let ctx = CustomWidgetContext {
                    state,
                    app_area: frame.area(),
                    events: channel.0,
                };

                view_select.render_ref(frame.area(), frame.buffer_mut(), &ctx);
            })
            .unwrap();

        assert_snapshot!(terminal.backend());
        tear_down(conf_path);
    }
}
