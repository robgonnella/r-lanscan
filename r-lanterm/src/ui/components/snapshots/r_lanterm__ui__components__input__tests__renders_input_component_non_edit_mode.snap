---
source: r-lanterm/src/ui/components/input.rs
expression: terminal.backend()
---
"Test: value                                                                     "
"                                                                                "
"                                                                                "
